### ISSUE TITLE
[OJS] Automated Submission Review Reminders are sent with the url and signature of another journal

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2274]
Hi,

Yesterday two submission review reminders were sent with the url and the signature of another journal.

The situation was this:

In journal A there were two overdue review assignments. Automatic reminders were turned off.
In juornal B there are no review assignments yet, but the automatic reminders are turned on.

Yesterday, the automatic reminders in journal B were triggered and the system wrote two reminder letters concerning the overdue assignments in journal A and sent them to the reviewers of journal A.

The letters included the initials of journal B in the subject field, the URLs were leading to journal B but had the submission id from journal A. Also, the signature and the sender was the editor of Journal B. All the rest of the information, receiver, submission title, abstract etc. came from journal A.

There were two similar cases and both review assignments were made already in OJS2.
Now using 3.0.2.

**Comment by ajnyga** [comment]
I guess that somehow the context used here https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L55 is not the context of the review assignment but the current context that has triggered the Acron plugin?

Although the context should be set correctly here: https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L156

I checked the database and there are no mistakes there. The submission context is what it should be and submission id in the review assignments is correct as well.

**Comment by ajnyga** [comment]
I checked from a database dump that `numDaysBeforeSubmitReminder` for journal A was set to `0`.
For journal B it is set to `1`

**Comment by ajnyga** [comment]
I can not figure out what has happened here. The only explanation I can think of is that for some reason this https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L156 has returned the wrong context for some reason. But that just does not seem possible.

**Comment by asmecher** [comment]
@ajnyga, are you using the acron plugin? If so, that might be worth investigating. That plugin piggy-backs onto normal user requests and performs scheduled tasks at the end of the request; since it's coming in on a user request, probably within a journal context, it's possible that the user's request is somehow getting used by the scheduled task e.g. to determine what journal to work with.

**Comment by ajnyga** [comment]
Hi @asmecher 

Yes the situation described above occured while using Acron. However, even after multiple attempts I have not been able to reproduce this. I am currently testing this by logging the reminders to the error log instead of actually sending them. So far everything seems to work ok.

As I mention above, the only logical explanation is that this https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L156 has returned the wrong context because `numDaysBeforeSubmitReminder` on line 159 has returned a value and therefore `$reminderType` is set to `REVIEW_REMIND_AUTO` in line 167 triggering line 177.

But I do not know how this `$context = $contextDao->getById($submission->getContextId())` could return a wrong context because the rest of the submission details are correct.

**Comment by ajnyga** [comment]
Ok, so yesterday my debugging script has caught a similar situation.

I will try to find out now what exactly happened, but the context from the submission and the context from context=>getId do not match!

**Comment by ajnyga** [comment]
Basically starting from https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L141 I have this:

```
foreach ($incompleteAssignments as $reviewAssignment) {
			// Avoid review assignments that a reminder exists for.
			if ($reviewAssignment->getDateReminded() !== null) continue;

			// Fetch the submission and the context.
			if ($submission == null || $submission->getId() != $reviewAssignment->getSubmissionId()) {
				unset($submission);
				$submission = $submissionDao->getById($reviewAssignment->getSubmissionId());
				// Avoid review assignments without submission in database.
				if (!$submission) continue;

				if ($submission->getStatus() != STATUS_QUEUED) continue;

				if ($context == null || $context->getId() != $submission->getContextId()) {
					unset($context);
					$context = $contextDao->getById($submission->getContextId());

					$inviteReminderDays = $context->getSetting('numDaysBeforeInviteReminder');
					$submitReminderDays = $context->getSetting('numDaysBeforeSubmitReminder');
					

					
				}
			}

			error_log(print_r("START submission: ", true));
			error_log(print_r($submission->getId(), true));

			
			$reminderType = false;
			if ($submitReminderDays>=1 && $reviewAssignment->getDateDue() != null) {
				
					error_log(print_r("Reminder: context, submission, remainderdays, datedue:", true));
					error_log(print_r($context->getId(), true));
					error_log(print_r($submission->getContextId(), true));
					error_log(print_r($submission->getId(), true));
					error_log(print_r($reviewAssignment->getSubmissionId(), true));
					error_log(print_r($submitReminderDays, true));
					error_log(print_r($reviewAssignment->getDateDue(), true));
				
				$checkDate = strtotime($reviewAssignment->getDateDue());
				if (time() - $checkDate > 60 * 60 * 24 * $submitReminderDays) {
					$reminderType = REVIEW_REMIND_AUTO;
					
					error_log(print_r($reminderType, true));
				}
			}
			if ($inviteReminderDays>=1 && $reviewAssignment->getDateConfirmed() == null) {
				$checkDate = strtotime($reviewAssignment->getDateResponseDue());
				if (time() - $checkDate > 60 * 60 * 24 * $inviteReminderDays) {
					$reminderType = REVIEW_REQUEST_REMIND_AUTO;
				}
			}
			
			error_log(print_r("END submission", true));

			
			if ($reminderType) $this->sendReminder ($reviewAssignment, $submission, $context, $reminderType);
		}
```

And yesterday in one case (there were several cases with no problems) that returned:

[Mon Feb 27 16:52:47 2017] [error] [client ] START submission: 
[Mon Feb 27 16:52:47 2017] [error] [client ] 60075
[Mon Feb 27 16:52:47 2017] [error] [client ] Reminder: context, submission, remainderdays, datedue:
[Mon Feb 27 16:52:47 2017] [error] [client ] 71
[Mon Feb 27 16:52:47 2017] [error] [client ] 82
[Mon Feb 27 16:52:47 2017] [error] [client ] 60075
[Mon Feb 27 16:52:47 2017] [error] [client ] 60075
[Mon Feb 27 16:52:47 2017] [error] [client ] 1
[Mon Feb 27 16:52:47 2017] [error] [client ] 2015-02-12 00:00:00
[Mon Feb 27 16:52:47 2017] [error] [client ] REVIEW_REMIND_AUTO
[Mon Feb 27 16:52:47 2017] [error] [client ] END submission

Meaning that `$context->getId()` and `$submission->getContextId()` are returning a different value!

I do no have the time to look into this right now, but what looks odd is the value for `$reviewAssignment->getDateDue()`.

**Comment by ajnyga** [comment]
A few details to this case.
- The submission 60075 is from journal 82.
- Journal 82 has both `numDaysBeforeSubmitReminder` and `numDaysBeforeInviteReminder` set to `0`
- Journal 71 has both `numDaysBeforeSubmitReminder` and `numDaysBeforeInviteReminder` set to `1`

- Submission 60075 is visible also in earlier debug logs but before yesterday `if ($submitReminderDays>=1 && $reviewAssignment->getDateDue() != null)` has not been triggered, meaning that an earlier print out looks like this:
```
[Sun Feb 26 16:52:45 2017] [error] [client ] START submission: 
[Sun Feb 26 16:52:45 2017] [error] [client ] 60075
[Sun Feb 26 16:52:45 2017] [error] [client ] END submission
```

Weird, weird, weird

**Comment by ajnyga** [comment]
Ok, so I think that I know what is happening here, I will post the solution in a while.

**Comment by ajnyga** [comment]
$reviewAssignmentDao->getIncompleteReviewAssignments() is returning these review assignment id's in this order:
```
4032 (submission 59705 journal 71) <- this is the context used in the email
3915 (submission 60064 journal 37)
3903 (submission 60064 journal 37)
3842 (submission 60075 journal 82)
3843 (submission 60075 journal 82) <- this triggers the email

```
With 3915 the loop ends here https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L152
With 3903 the loop ends here https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L143
With 3842 the loop ends here https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L152

With 4032 the `$context` is set to 71
3915, 3903 or 3842 do not get that far in the loop so that these would change the value for `$context`
However, 3842 does get so far that is changes the `$submission` to 60075.

So when 3843 starts the loop `$context` is set to 71 and `$submission` to 60075 although the submission 60075 does not belong to that context!

Furthermore, in https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L146 `if ($submission == null || $submission->getId() != $reviewAssignment->getSubmissionId())` returns false because `$submission->getId()` returns 60075 and `$reviewAssignment->getSubmissionId()` returns 60075. 

So basically the script skips everything between https://github.com/pkp/pkp-lib/blob/ojs-stable-3_0_2/classes/task/ReviewReminder.inc.php#L146-L161 and proceeds to sending a faulty email

**Comment by ajnyga** [comment]
There are probably several ways of dealing with this, but this is one working solution: https://github.com/pkp/pkp-lib/pull/2317

@asmecher @bozana what do you think? Is it a bad solution to reload the submission object for every reviewassignment?

**Comment by ajnyga** [comment]
On second thought, this could be better than the solution in the pr: 

https://gist.github.com/ajnyga/24964fb0a595a16917bd0e6bee66e42d

**Comment by ajnyga** [comment]
Another issue related to this.

This code https://gist.github.com/ajnyga/24964fb0a595a16917bd0e6bee66e42d does fix the problem I described above with the whole context changing in some cases, explained here: https://github.com/pkp/pkp-lib/issues/2274#issuecomment-283079770

However, there is another related issue. If you have this `{$contextUrl}` in a mail template it can be pretty much anything depending on the journal that is loaded when Acron runs. 
See: https://github.com/pkp/pkp-lib/blob/master/classes/mail/MailTemplate.inc.php#L160

I think that should be `'contextUrl' => $this->context->getPath()`, or what do you think?

@bozana or @asmecher would you look at these changes? I do think that this is fairly important.

**Comment by bozana** [comment]
PR: 
ojs master: https://github.com/pkp/pkp-lib/pull/2357

**Comment by bozana** [comment]
The PR above is merged, I will cherry-pick it to ojs-stable-3_0_2...

**Comment by bozana** [comment]
Everything merged, so can this issue be closed?

**Comment by ajnyga** [comment]
Great thank you @bozana , I will close this now!
